docs: DOC-2043 add network overlay preparation guide (#7562)
spectrocloud/librarium: docs/docs-content/deployment-modes/agent-mode/overlay-preparation.md

---
sidebar_label: "Configure networkd to Prepare Host for Overlay Network"
title: "Configure networkd to Prepare Host for Overlay Network"
description: "Learn how to configure your host to prepare it for enabling overlay network during cluster creation."
hide_table_of_contents: false
toc_max_heading_level: 2
sidebar_position: 10
tags: ["edge", "agent mode"]
---

Kubernetes clusters expect stable IP addresses that are not always possible in the network conditions of Edge locations.

> Palette Edge allows you to enable an overlay network to ensure the cluster has stable IP addresses even if the

Reviewer: Perhaps, it's worth adding a link for Network overlay https://docs.spectrocloud.com/clusters/edge/networking/vxlan-overlay/ ?

## Prerequisites

- This guide assumes your host uses Ubuntu as its OS. If you use a different OS distribution, you may need to find the
  equivalent commands that can accomplish the same goals in your OS.

- A Linux host that meets
  [the minimum requirements for agent mode deployments](architecture.md#minimum-device-requirements).

## Configure networkd

You can configure your host to make it overlay-ready either by issuing commands in the terminal to install the
prerequisite packages and configuring them, or by using a declarative block in the installer configuration user data to
run those commands automatically during agent installation.

<Tabs groupId="method">

<TabItem value="Manual Configuration in Command-line">

1. Issue the following commands to enable and restart `systemd-resolved`, and create a symlink between
   `/run/systemd/resolve/resolv.conf` and `/etc/resolv.conf`. This ensures that `systemd-resolved` starts managing DNS
   resolution for your host immediately.

   ```bash
   sudo systemctl enable systemd-resolved
   sudo systemctl restart systemd-resolved
   sudo ln -sf /run/systemd/resolve/resolv.conf /etc/resolv.conf
   ```

> 2. Create a backup of your current DNS configurations and remove them. The following commands are an example of how to

Reviewer 2: This is backup of netplan config for the nerwork interfaces and not just DNS.

   naming schemes, change this pattern so that the configuration matches.

   ```bash {3}
   cat > /etc/systemd/network/20-dhcp.network << EOF
   [Match]
   Name=en*
   [Network]
   DHCP=yes
   [DHCP]
   ClientIdentifier=mac
   EOF
   ```

5. Issue the following commands to enable and restart `systemd-networkd`. If you are connected to your host with an SSH
   connection, this will terminate the session because your host IP will change.

   ```bash
   sudo systemctl enable systemd-networkd
   sudo systemctl restart systemd-networkd
   ```

6. You have now configured your host to use `systemd-networkd` for DNS resolution and your cluster with this host will
   now support overlay network.

</TabItem>

<TabItem value="User Data Block During Agent Installation">

The installer configuration user data can declaratively configure your host to perform specified actions during
`cloud-init` stages. You can use one user data block to configure the same DNS settings across different hosts. We break
down the configuration step by step to explain each block. However, if you would like to copy the entire configuration,

> expand the following details block to copy the entire configuration.

Reviewer: ```suggestion
expand the following details block.
```
To avoid repeating "to copy the entire configuration"

               ClientIdentifier=mac
   ```

5. Enable and restart the `systemd-networkd` service so it takes over interface management.

   ```yaml
   #cloud-config
   stages:
     initramfs:
       - name: "Enable and restart systemd-networkd"
         commands:
           - systemctl enable systemd-networkd.service
           - systemctl restart systemd-networkd.service
   ```

6. Enable and restart `systemd-resolved` to handle DNS resolution consistently with `systemd`.

   ```yaml
   #cloud-config
   stages:
     initramfs:
       - name: "Enable and restart systemd-resolved"
         commands:
           - systemctl enable systemd-resolved.service
           - systemctl restart systemd-resolved.service
   ```

7. Ensure `/etc/resolv.conf` points to the DNS configuration managed by `systemd-resolved`.

   ```yaml
   #cloud-config
   stages:
     initramfs:
       - name: "Link systemd-resolved DNS to /etc/resolv.conf"
         if: '[ ! -L /etc/resolv.conf ] || [ "$(readlink -f /etc/resolv.conf)" != "/run/systemd/resolve/resolv.conf" ]'
         commands:
           - ln -sf /run/systemd/resolve/resolv.conf /etc/resolv.conf
   ```

</TabItem>

</Tabs>

## Validate

> Use the following steps to ensure that your DNS is now managed by `systemd-networkd`.

Reviewer 2: Maybe we should say "Use the following steps to ensure that your Network is now managed by `systemd-networkd`."

1. Issue the following command to retrieve a list of your network interfaces.

   ```bash
   networkctl
   ```

   `networkctl` is installed as part of the `systemd` package. If `networkctl` is not installed, then it is unlikely
   that your network is being managed by `systemd-networkd`.

>    Look for entries similar to the following.

Reviewer: I think it would be great to add what exactly should be similar, so it's more clear what to look for.

   ```bash
   IDX LINK  TYPE     OPERATIONAL SETUP
   1 lo    loopback carrier     unmanaged
   2 enp0s3 ether    routable    configured

>    ```

Reviewer: Is it possible to adjust the spaces in the snippet so the rows are more distinguishable from each other?

> 2. Issue the following command for non-loopback interfaces.

Reviewer: ```suggestion
2. Issue the following command for interfaces whose `TYPE` is not `loopback`.
```
I think this would make it more clear

   ```bash
   networkctl status enp0s3
   ```

   Look for the following to confirm that your expected interfaces are managed by `systemd-networkd`.

   ```bash
   Managed by: systemd-networkd
   ```

   This confirms that your host is using `systemd-networkd` for interface management.

3. Issue the following command to confirm the `systemd-resolved` service is active.

   ```bash
   systemctl is-active systemd-resolved
   ```

4. Issue the following command to check if `/etc/resolv.conf` is symlinked to `resolved`.

   ```bash
   readlink -f /etc/resolv.conf
   ```

   The expected output should be the following. This output, together with the service itself being active, confirms

>    that your host uses `systemd-resolved` for DNS management.

Reviewer 2: that your host uses `systemd-resolved` for Network and DNS management.

   ```text
   /run/systemd/resolve/resolv.conf
   ```

## Next Steps

The overlay network can only be enabled during cluster creation. Once a cluster is already created, you cannot enable
the overlay network without first tearing down the cluster. Refer to

> [Enable Overlay Network - Centrally Managed Cluster](../../clusters/edge/networking/vxlan-overlay.md) to learn more

Reviewer: I couldn't find "Centrally Managed Cluster" content on the linked page

Author: That page is about centrally managed clusters because the page describes how to enable overlay from palette, not local UI. But yeah I will add a sentence to clarify

about whether or not you should enable network overlay.

> Refer to [Create Local Cluster](../../clusters/edge/local-ui/cluster-management/create-cluster.md) to learn about how to
> create a local cluster, during which you can decide whether or not to enable the overlay network.

Reviewer: ```suggestion
Refer to [Create Local Cluster](../../clusters/edge/local-ui/cluster-management/create-cluster.md) to learn how to create a local cluster and choose whether to enable the overlay network.
```
